datafusion/core/src/physical_plan/stream.rs

```diff
@@ -17,43 +17,178 @@
+        F: FnOnce(),
+        F: Send + 'static,
+    {
+        self.join_set.spawn_blocking(f);
+    }
+
+    /// runs the input_partition of the `input` ExecutionPlan on the
+    /// tokio threadpool and writes its outputs to this stream
+    pub(crate) fn run_input(
+        &mut self,
+        input: Arc<dyn ExecutionPlan>,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) {
+        let output = self.tx();
+
+        self.spawn(async move {
+            let mut stream = match input.execute(partition, context) {
+                Err(e) => {
+                    // If send fails, plan being torn down,
+                    // there is no place to send the error.
+                    output.send(Err(e)).await.ok();
+                    debug!(
+                        "Stopping execution: error executing input: {}",
+                        displayable(input.as_ref()).one_line()
+                    );
+                    return;
+                }
+                Ok(stream) => stream,
+            };
+
+            while let Some(item) = stream.next().await {
+                // If send fails, plan being torn down,
+                // there is no place to send the error.
+                if output.send(item).await.is_err() {
+                    debug!(
+                        "Stopping execution: output is gone, plan cancelling: {}",
+                        displayable(input.as_ref()).one_line()
+                    );
+                    return;
+                }
+            }
+        });
+    }
 
-    #[allow(dead_code)]
-    drop_helper: AbortOnDropSingle<()>,
+    /// Create a stream of all `RecordBatch`es written to `tx`
+    pub fn build(self) -> SendableRecordBatchStream {
+        let Self {
+            tx,
+            rx,
+            schema,
+            mut join_set,
+        } = self;
+
+        // don't need tx
+        drop(tx);
+
+        // future that checks the result of the join set
+        let check = async move {
+            while let Some(result) = join_set.join_next().await {
+                match result {
+                    Ok(()) => continue, // nothing to report
+                    // This means a tokio task error, likely a panic
+                    Err(e) => {
+                        if e.is_panic() {
+                            // resume on the main thread
+                            std::panic::resume_unwind(e.into_panic());
+                        } else {
+                            return Some(Err(DataFusionError::Execution(format!(
+                                "Task error: {e}"
+                            ))));
+                        }
+                    }
+                }
+            }
+            None
+        };
+
+        let check_stream = futures::stream::once(check)
+            // unwrap Option / only return the error
+            .filter_map(|item| async move { item });
+
+        let inner = ReceiverStream::new(rx).chain(check_stream).boxed();
```

Comment:
This change looks much better than mine. Wondering if it can be improved further and if it makes sense at all.

With this implementation, the panics will be propagated only after all input (from other partitions is consumed). Probably fine, as this shouldn't happen during normal operation and is more of a correctness check. Also, the `check` future will not make any progress up until all the inputs are exhausted. Shouldn't be much work, fine for it to be sequential.

As an alternative, what if we build a "supervisor" task (`tokio::spawn`) which is launched to do all that work, and then in the `check_stream` we just check the `JoinHandle` of the "supervisor" task? This way the supervisor task will be able to make progress concurrently and panic/report errors early.

Thought about this after looking at `RepartitionStream` which would need something similar (supervisor) to get task failures and then multiplex them to all the output partitions. Then, all "output streams" would only have to ensure that the supervisor didn't die. Currently, in `RepartitionStream` there are `|output partitions|` "supervisors" (`wait_for_task`) which aren't checked for success either. Wondering if it could fail at all though (https://github.com/tokio-rs/tokio/discussions/5744).

Reply:
https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.take_while is possibly a better way to formulate this

Reply:
> With this implementation, the panics will be propagated only after all input

This is my worry as well. I think you could move the check future into another task (that holds the join set and is also aborted on drop, like a two-level join set) and that sends the error to `tx`.

Reply:
I am working on tests for this behavior

Reply:
I believe I fixed this in  https://github.com/apache/arrow-datafusion/pull/6507/commits/b1a817ce8e6c4eb210b40030f345ffb246184ac3

After trying several other approaches, I found  https://docs.rs/tokio-stream/latest/tokio_stream/trait.StreamExt.html#method.merge which did exactly what I wanted 💯 

It is tested in `record_batch_receiver_stream_propagates_panics_early_shutdown`